libs/framework/include/celix/dm/Component_Impl.h

```diff
@@ -63,7 +88,7 @@ Component<T>& Component<T>::addInterfaceWithName(const std::string &serviceName,
 
 template<class T>
 template<class I>
-Component<T>& Component<T>::addInterface(const std::string &version, const Properties &properties) {
+inline Component<T>& Component<T>::addInterface(const std::string &version, const Properties &properties) {
```

Comment:
I'm generally not in favour of using the inline keyword, the compiler is free to ignore it and usually knows better than programmers anyway. Moreover, templated functions in headers are already inline by default.

Reply:
function are default inline if you implement them in the class. 
In my experience the inline is needed if you implement it in a "impl" header (at least for non template functions)

Reply:
Rather than experience, let's look at the assembly generated: https://godbolt.org/z/cYasbb

No difference. Inline does not do what you think, it is only useful when getting linkage issues, which is not the case here.

Reply:
Removing all inline statements from the C++ dep man implementation, leads a lot of errors like:
```
/usr/bin/ld: CMakeFiles/test_framework.dir/src/bundle_context_services_test.cpp.o: in function `std::unique_lock<std::mutex>::~unique_lock()':
/home/pep/workspace/celix/libs/framework/include/celix/dm/ServiceDependency_Impl.h:28: multiple definition of `celix::dm::BaseServiceDependency::runBuild()'; CMakeFiles/test_framework.dir/src/bundle_context_bundles_tests.cpp.o:/home/pep/workspace/celix/libs/framework/include/celix/dm/ServiceDependency_Impl.h:28: first defined here
```

I read the stackoverflow link, it does not cover everything and I have not yet found out exactly why this is needed (I would like to understand it). 
Closest I got is:
https://isocpp.org/wiki/faq/templates#separate-template-class-defn-from-decl

As far as I know defining function inside a class will make them implicitly inline:
https://en.cppreference.com/w/cpp/language/inline
```
A function defined entirely inside a class/struct/union definition, whether it's a member function or a non-member friend function, is implicitly an inline function. 
```
So defining them outside a class, but still in a header - as is done with the _Impl.h headers - require a inline statement to make them effectually the same. 
Btw I used the _Imp.h construction, because I really prefer to have clean declarations of functions, classes, etc without cluttering it with implementation details.

Reply:
:laughing: 

You've managed to use inline in both the situation where it is useless and exactly the one type of situation where it is totally necessary.

Useless because templated functions are inline by default:
```c++
template<typename T>
inline void DependencyManager::destroyComponent(Component<T> &component) {
    celix_dependencyManager_remove(cDepMan, component.cComponent());
}
```

Totally necessary (if you want to keep it in a header file) because otherwise the function gets defined multiple times:
```
inline void DependencyManager::clear() {
    celix_dependencyManager_removeAllComponents(cDepMan.get());
    components.clear();
}
```

So one could make an argument that the `start()`, `build()`, `clear()`, `stop()`, `getNrOfComponents()` etc functions should be put in a .cpp file with only their definitions in the header file, since that would reduce compilation times and also solve the linking problem. This would create a 'function' in assembly that would be called as you would expect.

But inline there causes the function to not exist as a function but rather the assembly of a function to get duplicated everywhere it would normally be called. Whether that actually speeds things up is debatable, but it also means that there are no duplicate functions when linking.

Reply:
Consider the following situation:

example.h
```c++
#pragma once
void someFunction() {
   printf("funcy!");
}
```

some_file.cpp:
```c++
#include <example.h>

void stuff() {
  someFunction();
}
```

some_other_file.cpp:
```c++
#include <example.h>

void other_stuff() {
  someFunction();
}
```


What the compiler does when it tries to compile some_file.cpp is it looks for example.h, finds it, simply pastes the code into the translation unit (still some_file.cpp) and sees that a function `someFunction` is not only defined but also implemented. It generates the `someFunction` assembly and outputs some_file.o.

Then the compiler see some_other_file.cpp and does the same thing. Looks for the header, sees a definition and an implementation, creates the assembly for `someFunction` and outputs it into some_other_file.o.

Now the linker is told to create some_executable from the two object files some_file.o and some_other_file.o. But he sees `someFunction` twice and does not know how to deal with it. Which one is correct? It doesn't know how to determine that and instead of guessing what you meant and creating a random executable, it gives you an error like the following:

```
/usr/bin/ld: CMakeFiles/dm_example_cxx_phase3.dir/src/Phase3BaseActivator.cc.o: in function `std::function<void (IPhase2*, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >&&)>::operator bool() const':
/home/oipo-unencrypted/Programming/celix-apache/libs/framework/include/celix/dm/ServiceDependency_Impl.h:240: multiple definition of `celix::dm::DependencyManager::clear()'; CMakeFiles/dm_example_cxx_phase3.dir/src/Phase3Activator.cc.o:/home/oipo-unencrypted/Programming/celix-apache/libs/framework/include/celix/dm/DependencyManager_Impl.h:77: first defined here
``` 

The key being "first defined here". It's telling you "which of these two functions do you want me to use? Remove the other one please".

You can either change the header to only have a definition like so:
example.h
```c++
#pragma once
void someFunction();
```
example.cpp
```
#include <example.h>
void someFunction() {
   printf("funcy!");
}
```

Which would only generate the assembly for `someFunction` when the compiler is compiling example.cpp into example.o and never for another translation unit.

Or you can add inline like so:
example.h
```c++
#pragma once
inline void someFunction() {
   printf("funcy!");
}
```

This will cause some_file.o to contain the function `stuff()` which doesn't do a function call, but instead simply puts the printf call in the `stuff` function and does the same for `some_stuff` in some_other_file.o. The function `someFunction` simply "disappears" from the generated assembly.

Reply:
In short: 
1) please remove inline from all templated functions.
2) two options
  a) keep inline for the non-templated functions, solves the double implementation issue.
  b) create a DependencyManager_Impl.cc and move the non-templated function implementations there, have the impl header only contain templated functions. Faster compilation times and no confusing inline anywhere.

Reply:
Thanks.

I removed the inline only for the template functions and that indeed works. I was under the impressions that the inline was also explicitly needed for template functions. 

For now, I want to keep the non template function inline. This will keep the celix/dm functionality header only, part of the celix_framework, but without linking the celix_framework to C++ (libstd++).

Reply:
:+1:

Reply:
Thanks for bearing with my rants.